### PR TITLE
Custom init containers and init scripts

### PR DESCRIPTION
**Pull request by almahmoud**
Re: https://github.com/CloudVE/galaxy-kubernetes/issues/30

Adds the possibility to specify full init containers, or simply inject `sh` scripts that will be run through a pre-made init container that uses the same image as the main galaxy containers, and mounts the persistent volume
I included a commented out example in `values-cvmfs`. One thing to note is that the initscripts container will run (echoing one line) even if no scripts are specified rn. I can however change it to either use the same init container that we're using, just thought it should be isolated ideally.

### CODE REVIEW

File: galaxy/templates/deployment-job.yaml
Comment (nuwang) [inline galaxy/templates/deployment-job.yaml]: Most other charts seem to map initScripts as configMaps (also postgresql and keycloak). I think this makes sense? https://github.com/helm/charts/search?q=initscripts&unscoped_q=initscripts

File: galaxy/values-cvmfs.yaml
Comment (nuwang) [inline galaxy/values-cvmfs.yaml]: We should probably call this extraInitContainers: https://github.com/helm/charts/search?q=extraInitContainers&unscoped_q=extraInitContainers

File: galaxy/templates/deployment-web.yaml
Comment (nuwang) [inline galaxy/templates/deployment-web.yaml]: Can we include the entire fragment like this?
https://github.com/helm/charts/blob/e35e046cf78c89d800362dd97fbd761d2c892a70/stable/keycloak/templates/statefulset.yaml#L62

File: galaxy/values-cvmfs.yaml
Comment (nuwang) [inline galaxy/values-cvmfs.yaml]: Should probably be extraInitScripts.

**Review by nuwang — CHANGES_REQUESTED** [review]

File: galaxy/templates/deployment-web.yaml
Comment (almahmoud) [reply galaxy/templates/deployment-web.yaml]: We can if we remove the `applyTo[Web|Pod]` booleans. But if we want the option to choose which deployment/pod they should be run in, we have to extract the relevant values to ignore those booleans (might be able to still do it more directly if we do a `range` loop and ignore the two keys specifically, maybe? Is that more desirable though?

File: galaxy/values-cvmfs.yaml
Comment (almahmoud) [reply galaxy/values-cvmfs.yaml]: I agree. Will make the change.

File: galaxy/values-cvmfs.yaml
Comment (almahmoud) [reply galaxy/values-cvmfs.yaml]: Can change this too

File: galaxy/templates/deployment-job.yaml
Comment (almahmoud) [reply galaxy/templates/deployment-job.yaml]: They are mapped as configmaps to inject into the containers, but run through an init container.
Are you saying we should just map them and not run them, or run them in the main container before the command? (that would make their effects more persistent imo, as opposed to how it is now where they are only persistent if they change something on the persistent disk)

File: galaxy/values-cvmfs.yaml
Comment (almahmoud) [reply galaxy/values-cvmfs.yaml]: Although looking at it, it seems that other charts are using `initScripts` not `extraInitScripts` I think

File: galaxy/templates/deployment-job.yaml
Comment (almahmoud) [reply galaxy/templates/deployment-job.yaml]: Changed to use configmaps (or secrets) to inject, and run in the main containers before the command. Only `sh` scripts supported for now.

File: galaxy/values-cvmfs.yaml
Comment (almahmoud) [reply galaxy/values-cvmfs.yaml]: Changed!

File: galaxy/templates/deployment-web.yaml
Comment (almahmoud) [reply galaxy/templates/deployment-web.yaml]: Changed to put them as full init containers in the values files with the special two keys `applyTo[Web|Pod]` that get `unset` in the deployment manifests before rendering. Contents still go through `tpl`.

File: galaxy/values-cvmfs.yaml
Comment (almahmoud) [reply galaxy/values-cvmfs.yaml]: Kept at `initScripts` as seen more commonly imo, although open to changing it if you think it's better!

**Comment by almahmoud** [comment]
I think most if not all the changes are done, and tested and works as expected! Ready for a new review when someone has time.

**Comment by almahmoud** [comment]
P.S. Leaving this PR open until everything gets approved to preserve the conversations since started before the shift to `galaxyproject` repo, but can switch it to merge directly there if that's desirable.

File: galaxy/templates/_helpers.tpl
Comment (nuwang) [inline galaxy/templates/_helpers.tpl:96]: Just thinking about this, should `shed_tool_conf.xml` also be in the mutable directory instead?

File: galaxy/templates/deployment-job.yaml
Comment (nuwang) [inline galaxy/templates/deployment-job.yaml]: since we are already passing this through `sh -c`, the second `sh` seems unnecessary?

File: galaxy/templates/deployment-job.yaml
Comment (nuwang) [inline galaxy/templates/deployment-job.yaml:51]: This is great!

File: galaxy/values.yaml
Comment (nuwang) [inline galaxy/values.yaml]: This is `initScripts` instead of `extraInitScripts` but it's `extraInitContainers` instead of `initContainers`? Should they be the same?

File: galaxy/templates/configmap-scripts.yaml
Comment (nuwang) [inline galaxy/templates/configmap-scripts.yaml:6]: I was thinking that we could use one configMap for all init containers, instead of one configMap per script, just to reduce the number of artefacts?

File: galaxy/templates/deployment-web.yaml
Comment (nuwang) [inline galaxy/templates/deployment-web.yaml]: This section is duplicated between web and job handlers. Should we consider moving the whole thing to a template to avoid repetition? (In fact, we have a fair bit of duplicate logic between the two, and I'm just raising this here as a general design issue)

**Review by nuwang — COMMENTED** [review]
This looks good to me. I've added some comments but I think it's ok to merge and address those later when time permits.

File: galaxy/templates/_helpers.tpl
Comment (almahmoud) [reply galaxy/templates/_helpers.tpl:96]: I agree that it probably should. Will make the change separately since this PR is getting bigger than initContainers already

File: galaxy/templates/deployment-job.yaml
Comment (almahmoud) [reply galaxy/templates/deployment-job.yaml:51]: It took a while to figure out, but I was so happy when it worked haha

File: galaxy/values.yaml
Comment (almahmoud) [reply galaxy/values.yaml]: My thought was that `extraInitContainers` indicate that there are already `initContainers` running by default and do disambiguate between the k8s native `initContainers` key and the one in `values`, so I think it almost has to be `extraInitContainers` imo. On the other hand, we are not injecting any `initScripts` by default right now, and many other charts use `initScripts`: https://github.com/helm/charts/search?q=initScripts&unscoped_q=initScripts vs https://github.com/helm/charts/search?q=extrainitscripts&unscoped_q=extrainitscripts vs https://github.com/search?q=extrainitscripts&type=Code . However, it does seem less parallel esthetically so I don't mind switching the `initScripts` to `extraInitScripts` if we don't care to follow other charts that much. Someone has to do it first ultimately haha.

File: galaxy/templates/configmap-scripts.yaml
Comment (almahmoud) [reply galaxy/templates/configmap-scripts.yaml:6]: I'm assuming you mean `initScripts` and that would work if they're all secrets or all configmaps, but right now there is the option to set `useSecret` to true on one but not all, which requires each of them being a separate object. The only way I can think of to avoid this is maybe loop through and set a boolean to true the first time a secret/configmap is found and share that section for all subsequent secrets/configmaps, so there would be a single configmap and a single secret. i'm not sure how easy it will be to do with helm templating but I can give it a shot later on. I guess for now one question would be if we want to keep `useSecret` per script or not, because that will be an interface change if we shift later, but the implementation can be shifted without changes to users.

File: galaxy/templates/deployment-web.yaml
Comment (almahmoud) [reply galaxy/templates/deployment-web.yaml]: Yes, there are many many duplicates, with only changes being `applyToJob` vs `applyToWeb` as the relevant boolean. One concern I have about adding more things to the helpers is readability of the templates.

File: galaxy/templates/deployment-job.yaml
Comment (almahmoud) [reply galaxy/templates/deployment-job.yaml]: Correct me if this doesn't make sense, but I think if we don't add the `sh` command to run each script it will execute it based on the first line (which would be desirable to allow other types of scripts if we want that), but will also require us changing the permissions on the file to make it executable: `sh: 1: /galaxy/server/scripts/initscripts/myFirstScript.sh: Permission denied`
It shouldn't be hard to do if that's what we want, so i guess the question is do we want that?

**Pull request merged**
